utils/lib/rr_graph/channel2.py

```diff
@@ -53,38 +52,93 @@ def __init__(self, tracks):
         self.trees = []
         self.tracks = sorted(tracks, key=lambda x: x[1] - x[0])
 
-    def _place_track(self, track):
-        """Add track to existing interval tree, if there is room (ie doesn't overlap)
-        If track won't fit in any existing tree, allocate a new tree for the track"""
-        for idx, tree in enumerate(self.trees):
-            if not tree.overlaps(track[0], track[1] + 1):
-                tree.add(
-                    Interval(begin=track[0], end=track[1] + 1, data=track[2])
-                )
-                return
+    def _start_track(self, track):
+        self.trees.append([track])
+
+    def _add_track_to_tree(self, track, idx=-1):
+        self.trees[idx].append(track)
 
-        self.trees.append(IntervalTree())
-        self.trees[-1].add(
-            Interval(begin=track[0], end=track[1] + 1, data=track[2])
-        )
+    def _verify_trees(self):
+        for tree in self.trees:
+            for a, b in zip(tree, tree[1:]):
+                assert a[1] <= b[0]
 
     def pack_tracks(self):
-        """pack all tracks"""
-        for track in self.tracks[::-1]:
-            self._place_track(track)
+        """pack all tracks
+
+        Algorithm:
+
+         1. Sort tracks by length, shortest tracks first.  Popping from back
+            of python lists is O(1).
+         2. Create stack for each starting values, inserting in length order.
+         3. Starting with the lowest starting value, greedly pack tracks
+            Algorithm weaknesses:
+             - Linear scan for lowest starting value
+             - Linear scan for packing
+
+            Both weaknesses are O(Number grid dim * Number of tracks) in
+            pathilogical case, however grid dimensions tend to be fairly small,
+            (e.g. 50T is 150), so scans are practically fast.
+
+            If grid dimension size grows large, revisit how to find the lowest
+            starting value and next bucket pack.  Relevant operation is given
+            coordinate, find next largest non-empty bucket.
+         3a. Pop largest track from smallest starting value, creating a new
```

Comment:
Can you make these a proper nested list?

https://github.com/ralsina/rst-cheatsheet/blob/master/rst-cheatsheet.rst#lists

Reply:
This isn't RST, and it unclear how we would even render this as RST?

Reply:
Docstring formatting is RST. This is because they end up being parsed by docutils/sphinx.